src/hostapi/pulseaudio/pa_linux_pulseaudio_block.c

```diff
@@ -0,0 +1,186 @@
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR
+ * ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
+ * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+ * WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/*
+ * The text above constitutes the entire PortAudio license; however,
+ * the PortAudio community also makes the following non-binding requests:
+ *
+ * Any person wishing to distribute modifications to the Software is
+ * requested to send the modifications to the original developer so that
+ * they can be incorporated into the canonical version. It is also
+ * requested that these non-binding requests be included along with the
+ * license above.
+ */
+
+/** @file
+ @ingroup common_src
+
+ @brief PulseAudio implementation of support for a host API.
+
+ This host API implements PulseAudio support for portaudio
+ it has callbackmode and normal write mode support
+*/
+
+#include "pa_linux_pulseaudio_block_internal.h"
+#include <unistd.h>
+
+/*
+    As separate stream interfaces are used for blocking and callback
+    streams, the following functions can be guaranteed to only be called
+    for blocking streams.
+*/
+
+PaError PaPulseAudio_ReadStreamBlock( PaStream * s,
+                                      void *buffer,
+                                      unsigned long frames )
+{
+    PaPulseAudio_Stream *l_ptrStream = (PaPulseAudio_Stream *) s;
+    PaPulseAudio_HostApiRepresentation *l_ptrPulseAudioHostApi =
+        l_ptrStream->hostapi;
+    PaError l_iRet = 0;
+    size_t l_lReadable = 0;
+    uint8_t *l_ptrData = (uint8_t *) buffer;
+    long l_lLength = (frames * l_ptrStream->inputFrameSize);
+
+    pa_threaded_mainloop_lock( l_ptrStream->mainloop );
+    while( l_lLength > 0)
+    {
+        long l_read = PaUtil_ReadRingBuffer( &l_ptrStream->inputRing, l_ptrData,
+                                             l_lLength );
+        l_ptrData += l_read;
+        l_lLength -= l_read;
+        if( l_lLength > 0 )
+            pa_threaded_mainloop_wait( l_ptrStream->mainloop );
+    }
+    pa_threaded_mainloop_unlock( l_ptrStream->mainloop );
+    return paNoError;
+}
+
+
+PaError PaPulseAudio_WriteStreamBlock( PaStream * s,
+                                       const void *buffer,
+                                       unsigned long frames )
+{
+    PaPulseAudio_Stream *l_ptrStream = (PaPulseAudio_Stream *) s;
+    PaPulseAudio_HostApiRepresentation *l_ptrPulseAudioHostApi =
+        l_ptrStream->hostapi;
+    int l_iRet = 0;
+    size_t l_lWritable = 0;
+    uint8_t *l_ptrData = (uint8_t *) buffer;
+    long l_lLength = (frames * l_ptrStream->outputFrameSize);
+    pa_operation *l_ptrOperation = NULL;
+
+    PaUtil_BeginCpuLoadMeasurement( &l_ptrStream->cpuLoadMeasurer );
+
+    while( l_lLength > 0)
```

Comment:
This loop be becomes an invite loop if something goes wrong.
It think we need a reasonable tonight or such.

Reply:
Now it shouldn't anymore